File: cli/azd/extensions/microsoft.azd.extensions/internal/resources/languages/dotnet/commands/ListenCommand.cs
using System.CommandLine;
using Microsoft.Azd;
using Spectre.Console;

public class ListenCommand
{
    private readonly AzdClient _azdClient;

    public ListenCommand(AzdClient azdClient)
    {
        _azdClient = azdClient;
    }

    public Command Build()
    {
        var cmd = new Command("listen", "Starts the extension and listens for events.");

        cmd.SetHandler(async context =>
        {
            var cancellationToken = context.GetCancellationToken();

            await using var eventManager = new EventManager(_azdClient);

            // === Project Event Handler: preprovision
            await eventManager.AddProjectEventHandlerAsync("preprovision", async args =>
            {
                for (int i = 1; i <= 20; i++)
                {
                    AnsiConsole.MarkupLine($"[green]{i}.[/] Doing important work in C# extension...");
                    await Task.Delay(250, cancellationToken);
                }
            }, cancellationToken);

            // === Service Event Handler: prepackage
            await eventManager.AddServiceEventHandlerAsync("prepackage", async args =>
            {
                for (int i = 1; i <= 20; i++)
                {
                    AnsiConsole.MarkupLine($"[blue]{i}.[/] Doing important work in C# extension...");
                    await Task.Delay(250, cancellationToken);
                }
            }, options: null, cancellationToken);

            // === Start listening (blocking call)
            try
            {
                await eventManager.ReceiveAsync(cancellationToken);
            }
            catch (Exception ex)
            {
                AnsiConsole.MarkupLineInterpolated($"[red]Error while receiving events:[/] {ex.Message}");
            }
        });

        return cmd;
    }
}

File: cli/azd/extensions/microsoft.azd.extensions/internal/resources/languages/python/commands/listen_command.py
import asyncio
import logging
import sys
import os
import traceback

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
from azd_client import AzdClient
from event_manager import EventManager

class ListenCommand:
    def __init__(self, azd_client: AzdClient):
        self.azd_client = azd_client

    async def execute(self):
        logging.info("[ListenCommand] EventManager initialized")
        event_manager = EventManager(self.azd_client)
        # Project Event Handler: preprovision
        logging.info("[ListenCommand] Registering project.preprovision event handler")
        await event_manager.add_project_event_handler("preprovision", self.preprovision_handler)

        # Service Event Handler: prepackage
        logging.info("[ListenCommand] Registering service.prepackage event handler")
        await event_manager.add_service_event_handler("prepackage", self.prepackage_handler)

        # Start listening (blocking call)
        logging.info("[ListenCommand] Starting to listen for events...")
        await event_manager.receive()

    async def preprovision_handler(self, args):
        try:
            logging.info("[EventHandler] Starting project.preprovision work")
            # Ensure output is flushed immediately
            for i in range(1, 21):
                message = f"{i}. Doing important work in Python extension..."
                print(message)
                logging.info(f"[EventHandler] {message}")
                await asyncio.sleep(0.25)
            logging.info("[EventHandler] Finished project.preprovision work")
        except Exception as e:
            logging.error(f"[EventHandler] Error in preprovision_handler: {e}")
            traceback.print_exc()
            raise

    async def prepackage_handler(self, args):
        try:
            logging.info(f"[EventHandler] Starting service.prepackage work for service: {args.service}")
            # Ensure output is flushed immediately
            for i in range(1, 21):
                message = f"{i}. Doing important work in Python extension..."
                print(message)
                logging.info(f"[EventHandler] {message}")
                await asyncio.sleep(0.25)
            logging.info("[EventHandler] Finished service.prepackage work")
        except Exception as e:
            logging.error(f"[EventHandler] Error in prepackage_handler: {e}")
            traceback.print_exc()
            raise
